tests/v2/validation/rbac/rbac_cert_rotation_test.go

```diff
@@ -0,0 +1,151 @@
+	clusterName        string
+}
+
+func (rb *CertRotationRbacTestSuite) TearDownSuite() {
+	rb.session.Cleanup()
+}
+
+func (rb *CertRotationRbacTestSuite) SetupSuite() {
+	testSession := session.NewSession()
+	rb.session = testSession
+
+	rb.namespace = defaultNamespace
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(rb.T(), err)
+
+	rb.client = client
+	rb.clusterName = client.RancherConfig.ClusterName
+	logrus.Infof("Cluster is %v", rb.clusterName)
+	require.NotEmptyf(rb.T(), rb.clusterName, "Cluster name to install should be set")
+	clusterID, err := clusters.GetClusterIDByName(rb.client, rb.clusterName)
+	require.NoError(rb.T(), err, "Error getting cluster ID")
+	rb.cluster, err = rb.client.Management.Cluster.ByID(clusterID)
+	require.NoError(rb.T(), err)
+}
+
+func (rb *CertRotationRbacTestSuite) ValidateCertRotationCluster(role string) {
+
+	log.Infof("Rotating Certificates of the cluster as %v", role)
+	err := rotatecerts.RotateCertificates(rb.standardUserClient, rb.clusterName, rb.namespace)
+	switch role {
+	case roleOwner, restrictedAdmin:
+		require.NoError(rb.T(), err)
+
+	case roleMember, roleProjectOwner, roleProjectMember:
+		require.Error(rb.T(), err)
+		assert.Equal(rb.T(), "Resource type [provisioning.cattle.io.cluster] is not updatable", err.Error())
+	}
+
+	kubeProvisioningClient, err := rb.client.GetKubeAPIProvisioningClient()
+	require.NoError(rb.T(), err)
+
+	result, err := kubeProvisioningClient.Clusters(rb.namespace).Watch(context.TODO(), metav1.ListOptions{
+		FieldSelector:  "metadata.name=" + rb.clusterName,
+		TimeoutSeconds: &defaults.WatchTimeoutSeconds,
+	})
+	require.NoError(rb.T(), err)
+
+	checkFunc := clusters.IsProvisioningClusterReady
+
+	err = wait.WatchWait(result, checkFunc)
+	assert.NoError(rb.T(), err)
+}
+
+func (rb *CertRotationRbacTestSuite) TestCertRotationRbac() {
+	clusterID, err := clusters.GetClusterIDByName(rb.client, rb.clusterName)
+	logrus.Infof("Cluster is %v", clusterID)
+	require.NoError(rb.T(), err)
+	if !(strings.Contains(clusterID, "c-m-")) {
+		rb.T().Skip("Skipping tests since cluster is not of type - k3s or RKE2")
+	}
+	tests := []struct {
+		name   string
+		role   string
+		member string
+	}{
+		{"Cluster Member", roleMember, standardUser},
+		{"Project Owner", roleProjectOwner, standardUser},
+		{"Project Member", roleProjectMember, standardUser},
+		{"Restricted Admin", restrictedAdmin, restrictedAdmin},
+		{"Cluster Owner", roleOwner, standardUser},
+	}
+	for _, tt := range tests {
+		rb.Run("Set up User with Cluster Role "+tt.name, func() {
+			newUser, err := createUser(rb.client, tt.member)
+			require.NoError(rb.T(), err)
+			rb.standardUser = newUser
+			rb.T().Logf("Created user: %v", rb.standardUser.Username)
+			rb.standardUserClient, err = rb.client.AsUser(newUser)
+			require.NoError(rb.T(), err)
+
+			subSession := rb.session.NewSession()
+			defer subSession.Cleanup()
+
+			createProjectAsAdmin, err := createProject(rb.client, rb.cluster.ID)
```

Comment:
Just a thought. Can we remove project creation from the for loop and move it above for loop? so we are not creating a new project for Cluster Owner, cluster member and restricted admin as they are not utilizing the created project. By moving this above for loop, we are making sure to create a single project and make use of it for ProjectOwner and ProjectMember.

Or if we want a new project for PO and PM we can move this block to line 126..

Reply:
agreed